### PR TITLE
Enable TestInetEndPoint and TestInetLayerDNS

### LINKED ISSUE #329: After Issue #300 Is Addressed, Move TestInetLayerDNS from noinst_PROGRAMS Back to check_PROGRAMS
After GitHub issue #300 is addressed, move the Inet layer unit and functional test, *TestInetLayerDNS* from `noinst_PROGRAMS` back to `check_PROGRAMS` in *src/inet/tests/Makefile.am*.

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Missing TestInetEndPoint and TestInetLayerDNS for Inet layer

 #### Summary of Changes
Add TestInetEndPoint and TestInetLayerDNS
Add IPv6 support when bringing up docker

Fix #329

### CODE REVIEW

**Review by saurabhst — APPROVED** [review]

**Review by mspang — APPROVED** [review]

**Review by gerickson — CHANGES_REQUESTED** [review]
Why did we remove all of the static qualifiers from these methods and symbols? Most (all) shouldn't need to be visible outside of the test.

File: src/inet/tests/TestInetLayer.h [hunk 1]
```diff
@@ -33,7 +33,8 @@ int TestInetAddress(void);
 int TestInetBuffer(void);
 int TestInetErrorStr(void);
 int TestInetTimer(void);
-
+int TestInetEndPoint(int argc, char * argv[]);
```
Comment (gerickson) [inline src/inet/tests/TestInetLayer.h:36]: These don't follow the signature of those above. We don't have any use cases that pass arguments to these at present.

File: src/inet/tests/TestInetLayer.h [hunk 2]
```diff
@@ -33,7 +33,8 @@ int TestInetAddress(void);
...
+int TestInetLayerDNS(int argc, char * argv[]);
```
Comment (gerickson) [inline src/inet/tests/TestInetLayer.h:37]: These don't follow the signature of those above. We don't have any use cases that pass arguments to these at present.

**Review by gerickson — DISMISSED** [review]
At a high-level looks good; thanks for taking up the torch on these.

Two changes:

1. The removal of static qualifiers seems unnecessary. Generally, fewer globals are better.
2. We don't have any defined use cases for argc/argv test entry points and that doesn't fit the pattern prevalent elsewhere. Given that, keep these test entry points as `(void)`.

**Review by rwalker-apple — APPROVED** [review]
approved once @gerickson's comments are addressed

File: src/inet/tests/TestInetLayer.h (hunk 1 above)
Comment (yunhanw-google) [reply src/inet/tests/TestInetLayer.h]: Done, thank you

File: src/inet/tests/TestInetLayer.h (hunk 2 above)
Comment (yunhanw-google) [reply src/inet/tests/TestInetLayer.h]: Done, thank you

**Comment by yunhanw-google** [comment]
> At a high-level looks good; thanks for taking up the torch on these.
> 
> Two changes:
> 
> 1. The removal of static qualifiers seems unnecessary. Generally, fewer globals are better.
> 2. We don't have any defined use cases for argc/argv test entry points and that doesn't fit the pattern prevalent elsewhere. Given that, keep these test entry points as `(void)`.

Done, thank you

**Comment by yunhanw-google** [comment]
All comments have been addressed, please help merge this, thank you.

**Review by woody-apple — APPROVED** [review]

**Review by gerickson — COMMENTED** [review]
LGTM, thanks!

**Comment by bzbarsky-apple** [comment]
@yunhanw-google  was it intended that this added a bunch of time to `./gn_build.sh`?  

Also, the tests added here (which now run as part of `./gn_build.sh`) fail for me on the "respContainsIPv6Addrs" assertions, so I can no longer have a "successful" build.... Is that the intended behavior?  (Note: my ISP has no ipv6 routing support, in case that matters for these tests.)

In general, running tests that require internet connectivity, as part of a "build" phase seems really weird to me.

**Comment by yunhanw-google** [comment]
> @yunhanw-google was it intended that this added a bunch of time to `./gn_build.sh`?
> 
> Also, the tests added here (which now run as part of `./gn_build.sh`) fail for me on the "respContainsIPv6Addrs" assertions, so I can no longer have a "successful" build.... Is that the intended behavior? (Note: my ISP has no ipv6 routing support, in case that matters for these tests.)
> 
> In general, running tests that require internet connectivity, as part of a "build" phase seems really weird to me.

Hi, @bzbarsky-apple , What is your build environment? Are you running it insider docker, are you running it using mac? In this case, we need to validate inet address and dns using internet connectivity. @gerickson any suggestion here for the ISP without ipv6 support?

@mspang do you think if we can make tests including inet unit tests not part in gn_build.sh, or add option like gn_build.sh --enable-tests or we add the build flag for those tests needs internet connectivity or IPV6 support, in CI, we still enable those flags, at default, the flag is disabled.

**Comment by bzbarsky-apple** [comment]
I'm building outside docker, and yes, on Mac.

I understand why the tests are doing what they're doing, and I don't have a good suggestion for a different way to write the tests, short of running our own local DNS server as part of the test.  But I'd really like a way to build, in a supported way, without running those test.

**Comment by yunhanw-google** [comment]
@bzbarsky-apple thanks for your reply. Yes, fully agree that we need a way to build, in a supported way without running these tests.
@mspang  just disable those tests for mac.
https://github.com/project-chip/connectedhomeip/pull/2725

**Comment by gerickson** [comment]
> > @yunhanw-google was it intended that this added a bunch of time to `./gn_build.sh`?

These tests should generally be pretty quick. The only scenario I could imagine that take consequential time is if there are timeouts due to failures.

> > Also, the tests added here (which now run as part of `./gn_build.sh`) fail for me on the "respContainsIPv6Addrs" assertions, so I can no longer have a "successful" build.... Is that the intended behavior? (Note: my ISP has no ipv6 routing support, in case that matters for these tests.)

That shouldn't matter. As long as IPv6 is supported with a link-local address, that should be sufficient. If that isn't the case, then there's a bug that needs to be fixed.

> > In general, running tests that require internet connectivity, as part of a "build" phase seems really weird to me.

Agreed, we should generally not need Internet connectivity for our tests. The DNS tests may break that. We should try to get to a place where we can make those optional or where we can stand up a local DNS responder.

> Hi, @bzbarsky-apple , What is your build environment? Are you running it insider docker, are you running it using mac? In this case, we need to validate inet address and dns using internet connectivity. @gerickson any suggestion here for the ISP without ipv6 support?

See above.

> @mspang do you think if we can make tests including inet unit tests not part in gn_build.sh, or add option like gn_build.sh --enable-tests or we add the build flag for those tests needs internet connectivity or IPV6 support, in CI, we still enable those flags, at default, the flag is disabled.

The only thing we should need to disable, for now until we can make them optional or can devise a local responder, are DNS tests.

**Comment by bzbarsky-apple** [comment]
> The only scenario I could imagine that take consequential time is if there are timeouts due to failures.

Well, I am certainly running into that.  ;)

> As long as IPv6 is supported with a link-local address, that should be sufficient. If that isn't the case, then there's a bug that needs to be fixed.

The failures are in DNS resolution, like so:
```
../../src/inet/tests/TestInetLayerDNS.cpp:601: assertion failed: "respContainsIPv6Addrs"
```
and yes, link-local (and everything up to the gateway) supports IPv6, and `dig -type ALL` shows `AAAA` records when they exist.  I can file an issue about the failures; please just let me know!  Happy to provide more logging information if there's a way to get it; otherwise I can just paste my build log in there.

**Pull request merged**